homeassistant/helpers/llm.py

```diff
@@ -0,0 +1,240 @@
+"""Module to coordinate llm tools."""
+
+from __future__ import annotations
+
+from abc import abstractmethod
+from collections.abc import Callable, Iterable
+from dataclasses import dataclass
+from functools import cached_property
+import logging
+from types import FunctionType
+from typing import Any
+
+import voluptuous as vol
+from voluptuous_openapi import UNSUPPORTED, convert
+
+from homeassistant.core import Context, HomeAssistant, callback
+from homeassistant.exceptions import HomeAssistantError
+
+from . import area_registry as ar, config_validation as cv, floor_registry as fr, intent
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_KEY = "llm_tool"
+
+
+def default_custom_serializer(schema: Any) -> Any:
+    """Serialize additional types in OpenAPI-compatible format."""
+    from homeassistant.util.color import (  # pylint: disable=import-outside-toplevel
+        color_name_to_rgb,
+    )
+
+    if schema is cv.string:
+        return {"type": "string"}
+
+    if schema is cv.boolean:
+        return {"type": "boolean"}
+
+    if schema is color_name_to_rgb:
+        return {"type": "string"}
+
+    if isinstance(schema, cv.multi_select):
+        return {"enum": schema.options}
+
+    if isinstance(schema, FunctionType):
+        return {}
+
+    return UNSUPPORTED
+
+
+@dataclass(slots=True)
+class ToolInput:
+    """Tool input to be processed."""
+
+    tool_name: str
+    tool_args: dict[str, Any]
+    platform: str
+    context: Context | None
+    user_prompt: str | None
+    language: str | None
+    agent_id: str | None
+    conversation_id: str | None
+    device_id: str | None
+    assistant: str | None
+
+
+class Tool:
+    """LLM Tool base class."""
+
+    name: str
+    description: str
+    parameters: vol.Schema
+    custom_serializer: Callable[[Any], Any]
+
+    def __init__(
+        self,
+        name: str,
+        description: str,
+        parameters: vol.Schema = vol.Schema({}),
+        custom_serializer: Callable[[Any], Any] = default_custom_serializer,
```

Comment:
I think serialization of the schema should be a responsibility of the calling integration and not the tool registry api.

That is, when I register a tool, i shouldn't really have opinions about how the tool schema is serialized in an OpenAPI-compatible format or not.

Or is this needed because a tool may register custom types that conversation agents wont know how to serialize?

Reply:
> Or is this needed because a tool may register custom types that conversation agents wont know how to serialize?

Exactly.

And this is also one of the reasons why I want to keep serialization into OpenAPI format as a standard feature.

Reply:
OK. Are there types you know of that can't easily be serialized right now?

Reply:
In the `default_custom_serializer` I tried to capture all validators that are currently used by intents, they can serve as an example. All built-in voluptuous validators and built-in types are handled by `voluptuous_openapi`, but any function or class can be a validator in voluptuous schema. Some of the `helpers.config_validation` are handled by `default_custom_serializer`, but not all of them, and I have no intention to maintain full compatibility with any possible validator in this module, and I want to give the tool developer some flexibility to properly represent the tool to the LLM if they support it.

Reply:
The `custom_serializer` was removed